File: compiler/src/main/java/com/aaroncoplan/waterfall/compiler/statements/ExpressionData.java
package com.aaroncoplan.waterfall.compiler.statements;

import com.aaroncoplan.waterfall.generated.WaterfallParser;

public class ExpressionData {

    public enum Kind {
        NULL_LITERAL,
        INT_LITERAL,
        DEC_LITERAL,
        STRING_LITERAL,
        IDENTIFIER,
        LAMBDA,
        BUNDLE,
        ARRAY,
        FUNCTION_CALL,
        BINARY_OP,
    }

    public final Kind kind;
    public final String literalText;            // NULL/INT/DEC/STRING/IDENTIFIER
    public final LambdaFunctionData lambda;     // LAMBDA only
    public final BundleLiteralData bundle;      // BUNDLE only
    public final ArrayLiteralData array;        // ARRAY only
    public final FunctionCallData functionCall; // FUNCTION_CALL only
    public final String op;                     // BINARY_OP only: "and" / "or" / "equals"
    public final ExpressionData left;           // BINARY_OP only
    public final ExpressionData right;          // BINARY_OP only

    public ExpressionData(String filePath, WaterfallParser.ExpressionContext ctx) {
        if (ctx.op != null) {
            this.kind = Kind.BINARY_OP;
            this.op = ctx.op.getText();
            this.left = new ExpressionData(filePath, ctx.left);
            this.right = new ExpressionData(filePath, ctx.right);
            this.literalText = null;
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
            return;
        }
        // No-op default for binary fields when this is a leaf.
        this.op = null;
        this.left = null;
        this.right = null;
        if (ctx.NULL() != null) {
            this.kind = Kind.NULL_LITERAL;
            this.literalText = ctx.NULL().getText();
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
        } else if (ctx.INT_LITERAL() != null) {
            this.kind = Kind.INT_LITERAL;
            this.literalText = ctx.INT_LITERAL().getText();
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
        } else if (ctx.DEC_LITERAL() != null) {
            this.kind = Kind.DEC_LITERAL;
            this.literalText = ctx.DEC_LITERAL().getText();
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
        } else if (ctx.STRING_LITERAL() != null) {
            this.kind = Kind.STRING_LITERAL;
            this.literalText = ctx.STRING_LITERAL().getText();
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
        } else if (ctx.lambdaFunction() != null) {
            this.kind = Kind.LAMBDA;
            this.literalText = null;
            this.lambda = new LambdaFunctionData(filePath, ctx.lambdaFunction());
            this.bundle = null; this.array = null; this.functionCall = null;
        } else if (ctx.bundleLiteral() != null) {
            this.kind = Kind.BUNDLE;
            this.literalText = null;
            this.bundle = new BundleLiteralData(filePath, ctx.bundleLiteral());
            this.lambda = null; this.array = null; this.functionCall = null;
        } else if (ctx.arrayLiteral() != null) {
            this.kind = Kind.ARRAY;
            this.literalText = null;
            this.array = new ArrayLiteralData(filePath, ctx.arrayLiteral());
            this.lambda = null; this.bundle = null; this.functionCall = null;
        } else if (ctx.functionCall() != null) {
            this.kind = Kind.FUNCTION_CALL;
            this.literalText = null;
            this.functionCall = new FunctionCallData(filePath, ctx.functionCall());
            this.lambda = null; this.bundle = null; this.array = null;
        } else if (ctx.ID() != null) {
            this.kind = Kind.IDENTIFIER;
            this.literalText = ctx.ID().getText();
            this.lambda = null; this.bundle = null; this.array = null; this.functionCall = null;
        } else {
            throw new RuntimeException("Unrecognized expression alternative");
        }
    }
}

File: compiler/src/main/kotlin/com/aaroncoplan/waterfall/compiler/statements/ExpressionData.kt
package com.aaroncoplan.waterfall.compiler.statements

import com.aaroncoplan.waterfall.generated.WaterfallParser

class ExpressionData(filePath: String, ctx: WaterfallParser.ExpressionContext) {

    enum class Kind {
        NULL_LITERAL,
        BOOL_LITERAL,
        INT_LITERAL,
        DEC_LITERAL,
        STRING_LITERAL,
        IDENTIFIER,
        LAMBDA,
        BUNDLE,
        ARRAY,
        FUNCTION_CALL,
        BINARY_OP,
        ARRAY_INDEX,
        CAST,
    }

    @JvmField val kind: Kind
    /** NULL/INT/DEC/STRING/IDENTIFIER/BOOL */
    @JvmField val literalText: String?
    /** LAMBDA only */
    @JvmField val lambda: LambdaFunctionData?
    /** BUNDLE only */
    @JvmField val bundle: BundleLiteralData?
    /** ARRAY only */
    @JvmField val array: ArrayLiteralData?
    /** FUNCTION_CALL only */
    @JvmField val functionCall: FunctionCallData?
    /** BINARY_OP only: "and" / "or" / "equals" / "+" / "-" / ... */
    @JvmField val op: String?
    /** BINARY_OP only */
    @JvmField val left: ExpressionData?
    /** BINARY_OP only */
    @JvmField val right: ExpressionData?
    /** ARRAY_INDEX only */
    @JvmField val arrayIndex: ArrayIndexData?
    /** CAST only — the target type name */
    @JvmField val castTargetType: String?
    /** CAST only */
    @JvmField val castOperand: ExpressionData?

    init {
        // Cast alt — `expression CASTAS castTarget=type`. Check before binary because
        // it's also left-recursive and we want to identify it precisely.
        if (ctx.castTarget != null) {
            kind = Kind.CAST
            castTargetType = ctx.castTarget.text
            castOperand = ExpressionData(filePath, ctx.expression(0))
            op = null; left = null; right = null
            literalText = null
            lambda = null; bundle = null; array = null
            functionCall = null; arrayIndex = null
        }
        // Binary op alternative — must be checked next because the `op` label is the
        // only thing distinguishing it (all other accessors return null in that case).
        else if (ctx.op != null) {
            kind = Kind.BINARY_OP
            op = ctx.op.text
            left = ExpressionData(filePath, ctx.left)
            right = ExpressionData(filePath, ctx.right)
            literalText = null
            lambda = null; bundle = null; array = null
            functionCall = null; arrayIndex = null
            castTargetType = null; castOperand = null
        }
        else {
            // Leaf / primary alternatives.
            var k: Kind? = null
            var text: String? = null
            var ld: LambdaFunctionData? = null
            var bd: BundleLiteralData? = null
            var ad: ArrayLiteralData? = null
            var fc: FunctionCallData? = null
            var ai: ArrayIndexData? = null
            when {
                ctx.NULL() != null         -> { k = Kind.NULL_LITERAL;   text = ctx.NULL().text }
                ctx.BOOL_LITERAL() != null -> { k = Kind.BOOL_LITERAL;   text = ctx.BOOL_LITERAL().text }
                ctx.INT_LITERAL() != null  -> { k = Kind.INT_LITERAL;    text = ctx.INT_LITERAL().text }
                ctx.DEC_LITERAL() != null  -> { k = Kind.DEC_LITERAL;    text = ctx.DEC_LITERAL().text }
                ctx.STRING_LITERAL() != null -> { k = Kind.STRING_LITERAL; text = ctx.STRING_LITERAL().text }
                ctx.lambdaFunction() != null -> { k = Kind.LAMBDA;       ld = LambdaFunctionData(filePath, ctx.lambdaFunction()) }
                ctx.bundleLiteral() != null  -> { k = Kind.BUNDLE;       bd = BundleLiteralData(filePath, ctx.bundleLiteral()) }
                ctx.arrayLiteral() != null   -> { k = Kind.ARRAY;        ad = ArrayLiteralData(filePath, ctx.arrayLiteral()) }
                ctx.functionCall() != null   -> { k = Kind.FUNCTION_CALL; fc = FunctionCallData(filePath, ctx.functionCall()) }
                ctx.arrayIndex() != null     -> { k = Kind.ARRAY_INDEX;  ai = ArrayIndexData(filePath, ctx.arrayIndex()) }
                ctx.ID() != null             -> { k = Kind.IDENTIFIER;   text = ctx.ID().text }
                else -> throw RuntimeException("Unrecognized expression alternative")
            }
            kind = k!!
            literalText = text
            lambda = ld
            bundle = bd
            array = ad
            functionCall = fc
            arrayIndex = ai
            op = null
            left = null
            right = null
            castTargetType = null
            castOperand = null
        }
    }
}
